h/browser/chrome/lib/browser-action.js

```diff
@@ -36,23 +36,88 @@
+     * @param {string} title The text to show in the tooltip.
+     * @param {string} badgeText The text to show on the badge.
+     */
+    function setTitleAndBadgeText(tabId, title, badgeText) {
+      chromeBrowserAction.setTitle({tabId: tabId, title: title});
+      chromeBrowserAction.setBadgeText({tabId: tabId, text: badgeText});
+    }
+
     /* Sets the active browser action appearance for the provided tab id. */
-    this.activate = function (tabId) {
+    this.activate = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.ACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis is active')});
+      setTitleIfNoBadge(tabId, _('Hypothesis is active'));
     };
 
     /* Sets the inactive browser action appearance for the provided tab id. */
-    this.deactivate = function (tabId) {
+    this.deactivate = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.INACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis is inactive')});
+      setTitleIfNoBadge(tabId, _('Hypothesis is inactive'));
     };
 
     /* Sets the errored browser action appearance for the provided tab id. */
-    this.error = function (tabId) {
+    this.error = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.INACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis has failed to load')});
-      chromeBrowserAction.setBadgeText({tabId: tabId, text: '!'});
+      setTitleAndBadgeText(tabId,  _('Hypothesis has failed to load'), '!');
+    };
+
+    /**
+     * Show the number of annotations of the current page in the badge.
+     * @method
+     * @param {integer} tabId The id of the current tab.
+     * @param {string} tabUrl The URL of the current tab.
+     * @param {string} serviceUrl The URL of the Hypothesis API.
+     */
+    this.updateBadge = function(tabId, tabUrl, serviceUrl) {
+      // Fetch the number of annotations of the current page from the server,
+      // and display it as a badge on the browser action button.
+      var xhr = new XMLHttpRequest();
+      xhr.onload = function() {
+        var total = 0;
+        total = JSON.parse(this.response).total;
+        if (total > 0) {
+          var totalString = total.toString();
+          if (total > 999) {
+            totalString = '999+';
+          }
+          chromeBrowserAction.getBadgeText({tabId: tabId}, function(text) {
+            if (!text) {
+              var title;
+              if (total === 1) {
+                title = _("There's 1 annotation on this page");
+              } else {
+                title = _('There are ' + totalString + ' annotations on ' +
+                          'this page');
+              }
+              setTitleAndBadgeText(tabId, title, totalString);
+            }
+          });
+        }
+      };
+
+      if (serviceUrl.charAt(serviceUrl.length - 1) !== '/') {
```

Comment:
This shouldn't be dealt with here. It should either be normalised in the settings module when loaded, or it should just be defined to omit the trailing slash.